Bluetooth: Audio: Media - writes to writeable object IDs

asbjornsabo opened the pull request:
Implement missing functions to do, and handle, writes to writable object IDs in the media control service.

Client functions to do the writes.
Shell functions corresponding to this
Proper handling of writes in the mediaplayer
BabbleSim tests

asbjornsabo committed:
Bluetooth: Audio: Media - MCC functions to write object IDs

Functions and callbacks to write the Current Track Object ID, the Next
Track Object ID and the Current Group Object ID.

Shell commands to exercise the write functions and receive the
results.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media - handle object ID writes in MPL

Actually handle the writes to object IDs in the correct way.

Current Track Object ID: Find that track and change to it, with all
required notifications.

Next Track Object ID: Find the next track, and set up for changing to
it at next "next track" operation.

Current Group Object ID: Find the group, and change to it, with all
required notification.

Rmove some code that was no longer needed after this.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

Handle next track write

Remove unused code

next-track fixes

asbjornsabo committed:
Bluetooth: Audio: Media - tests for object ID writes

Add BabbleSim tests for writes to object IDs.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/audio/mcc.c:
```diff
@@ -498,6 +498,28 @@ static uint8_t mcc_read_current_track_obj_id_cb(struct bt_conn *conn, uint8_t er
 }
 
+static void mcs_write_current_track_obj_id_cb(struct bt_conn *conn, uint8_t err,
+					      struct bt_gatt_write_params *params)
```
```suggestion
					      const struct bt_gatt_write_params *params)
```

asbjornsabo replied:
Changed as suggested, but had to change it back, as it broke the build.
The prototype in gatt does not have the const keyword.

Thalley commented on subsys/bluetooth/host/audio/mcc.c:
```diff
@@ -527,6 +549,28 @@ static uint8_t mcc_read_next_track_obj_id_cb(struct bt_conn *conn, uint8_t err,
 }
 
+static void mcs_write_next_track_obj_id_cb(struct bt_conn *conn, uint8_t err,
+					   struct bt_gatt_write_params *params)
```
```suggestion
					   const struct bt_gatt_write_params *params)
```

Thalley commented on subsys/bluetooth/host/audio/mcc.c:
```diff
@@ -554,6 +598,28 @@ static uint8_t mcc_read_current_group_obj_id_cb(struct bt_conn *conn, uint8_t er
 }
 
+static void mcs_write_current_group_obj_id_cb(struct bt_conn *conn, uint8_t err,
+					      struct bt_gatt_write_params *params)
```
```suggestion
					      const struct bt_gatt_write_params *params)
```

Thalley commented on subsys/bluetooth/host/audio/mcc.c:
```diff
@@ -1758,6 +1824,37 @@ int bt_mcc_read_current_track_obj_id(struct bt_conn *conn)
 	return err;
 }
 
+int bt_mcc_set_current_track_obj_id(struct bt_conn *conn, uint64_t obj_id)
```
```suggestion
int bt_mcc_current_track_obj_id_set(struct bt_conn *conn, uint64_t obj_id)
```

Thalley commented on subsys/bluetooth/host/audio/mcc.c:
```diff
@@ -1758,6 +1824,37 @@ int bt_mcc_read_current_track_obj_id(struct bt_conn *conn)
+	int err;
+
+	if (!conn) {
+		return -ENOTCONN;
```
If `conn == NULL`, then it doesn't really mean that is it not connected (that's a state in `conn`), but rather invalid argument. 
```suggestion
		return -EINVAL
```

asbjornsabo replied:
Fixed everywhere in mcc.c

Thalley commented on subsys/bluetooth/host/audio/mcc.c:
```diff
@@ -1758,6 +1824,37 @@ int bt_mcc_read_current_track_obj_id(struct bt_conn *conn)
+{
+	int err;
+
+	if (!conn) {
```
Prefer to use CHECKIF for static checks such as NULL checks so that it can easily be optimized later by the application if wanted. 
```suggestion
	CHECKIF(!conn) {
```

asbjornsabo replied:
Added to later task

Thalley commented on subsys/bluetooth/host/audio/mcc.c:
```diff
@@ -1758,6 +1824,37 @@ int bt_mcc_read_current_track_obj_id(struct bt_conn *conn)
+	cur_mcs_inst->write_params.handle = cur_mcs_inst->current_track_obj_id_handle;
+	cur_mcs_inst->write_params.func = mcs_write_current_track_obj_id_cb;
+
+	BT_HEXDUMP_DBG(cur_mcs_inst->write_params.data, BT_OTS_OBJ_ID_SIZE, "Object Id");
```
Consider using `bt_hex` instead, as they will also log the function name
```suggestion
	BT_DBG("Object ID: %s", bt_hex(cur_mcs_inst->write_params.data, BT_OTS_OBJ_ID_SIZE));
```

asbjornsabo replied:
Added to later task

Thalley commented on subsys/bluetooth/host/audio/mcc.c:
```diff
@@ -1785,6 +1882,37 @@ int bt_mcc_read_next_track_obj_id(struct bt_conn *conn)
 	return err;
 }
 
+int bt_mcc_set_next_track_obj_id(struct bt_conn *conn, uint64_t obj_id)
```
```suggestion
int bt_mcc_next_track_obj_id_set(struct bt_conn *conn, uint64_t obj_id)
```

Thalley replied:
Should we add a check for the `obj_id` value? It shall be greater than or equal to `0x100` or something, right? (applies multiple places)

asbjornsabo replied:
Added it to the three added functions.
Made upstream PR to get a #define into OTS

Thalley commented on subsys/bluetooth/host/audio/mcc.c:
```diff
@@ -1785,6 +1882,37 @@ int bt_mcc_read_next_track_obj_id(struct bt_conn *conn)
+{
+	int err;
+
+	if (!conn) {
+		return -ENOTCONN;
+	}
```
Ditto

Thalley commented on subsys/bluetooth/host/audio/mcc.c:
```diff
@@ -1785,6 +1882,37 @@ int bt_mcc_read_next_track_obj_id(struct bt_conn *conn)
+	cur_mcs_inst->write_params.handle = cur_mcs_inst->next_track_obj_id_handle;
+	cur_mcs_inst->write_params.func = mcs_write_next_track_obj_id_cb;
+
+	BT_HEXDUMP_DBG(cur_mcs_inst->write_params.data, BT_OTS_OBJ_ID_SIZE, "Object Id");
```
Ditto

asbjornsabo replied:
Added to later task

Thalley commented on subsys/bluetooth/host/audio/mcc.c:
```diff
@@ -1812,6 +1940,37 @@ int bt_mcc_read_current_group_obj_id(struct bt_conn *conn)
 	return err;
 }
 
+int bt_mcc_set_current_group_obj_id(struct bt_conn *conn, uint64_t obj_id)
```
```suggestion
int bt_mcc_current_group_obj_id_set(struct bt_conn *conn, uint64_t obj_id)
```

Thalley commented on subsys/bluetooth/host/audio/mcc.c:
```diff
@@ -1812,6 +1940,37 @@ int bt_mcc_read_current_group_obj_id(struct bt_conn *conn)
+{
+	int err;
+
+	if (!conn) {
+		return -ENOTCONN;
+	}
```
Ditto

Thalley commented on subsys/bluetooth/host/audio/mcc.c:
```diff
@@ -1812,6 +1940,37 @@ int bt_mcc_read_current_group_obj_id(struct bt_conn *conn)
+	cur_mcs_inst->write_params.handle = cur_mcs_inst->current_group_obj_id_handle;
+	cur_mcs_inst->write_params.func = mcs_write_current_group_obj_id_cb;
+
+	BT_HEXDUMP_DBG(cur_mcs_inst->write_params.data, BT_OTS_OBJ_ID_SIZE, "Object Id");
```
Ditto

Thalley commented on subsys/bluetooth/host/audio/mpl.c:
```diff
@@ -981,6 +967,31 @@ static bool do_next_track(struct mpl_mediaplayer_t *pl)
+ *
+ * Returns true if the _group_ has been changed, otherwise false
+ */
+static bool do_next_track_set(struct mpl_mediaplayer_t *pl)
```
```suggestion
static bool next_track_set(struct mpl_mediaplayer_t *pl)
```
or
```suggestion
static bool set_next_track(struct mpl_mediaplayer_t *pl)
```

Thalley replied:
Hmm, I'm getting confused about the naming. What's the difference between this function and `do_next_track`?

asbjornsabo replied:
Renamed the functions to do_next_track_normal_order() do_next_track_next_track_set().
The first one changes according to the preset order (goes to the current track's next track), the second one changes to the track that has been set as the next track.

Thalley replied:
Can you briefly explain why we need two different functions for changing to the next track?

asbjornsabo replied:
Well, they can of course be hidden under a common wrapper (or merged to one function).
But having them as two is clearer, as one of them is a special case, while the other follows the pattern of all the other "do_<prev|nest|first|last|goto>_<segment|track|group>.

asbjornsabo replied:
Also, they behave slightly differently with respect to the track position.

Thalley commented on subsys/bluetooth/host/audio/mpl.c:
```diff
@@ -2204,6 +2247,77 @@ void (*command_handlers[BT_MCS_MEDIA_STATE_LAST])(struct mpl_cmd command,
+ * If found, return pointers to the group of the track and the track
+ * otherwise, the pointers returned are NULL
+ */
+static void find_track_by_id(struct mpl_mediaplayer_t *pl, uint64_t id,
```
```suggestion
static void find_track_by_id(const struct mpl_mediaplayer_t *pl, uint64_t id,
```

Thalley commented on subsys/bluetooth/host/audio/mpl.c:
```diff
@@ -2204,6 +2247,77 @@ void (*command_handlers[BT_MCS_MEDIA_STATE_LAST])(struct mpl_cmd command,
+	struct mpl_group_t *tmp_group = pl->group;
+	struct mpl_track_t *tmp_track;
+
+	while (tmp_group->prev != NULL) {
+		tmp_group = tmp_group->prev;
+	}
```
This is to find the head, right? 

Could the head be stored in the `struct mpl_mediaplayer_t *pl` for direct and fast access?

asbjornsabo replied:
Yes.  But I consider that out of scope for this PR.

Thalley commented on subsys/bluetooth/host/audio/mpl.c:
```diff
@@ -2204,6 +2247,77 @@ void (*command_handlers[BT_MCS_MEDIA_STATE_LAST])(struct mpl_cmd command,
+	while (tmp_group != NULL) {
+		tmp_track = tmp_group->track;
+
+		while (tmp_track->prev != NULL) {
+			tmp_track = tmp_track->prev;
+		}
```
Likewise, would it make sense to store the head of the list of tracks in `struct mpl_group_t`?

Thalley commented on subsys/bluetooth/host/audio/mpl.c:
```diff
@@ -2204,6 +2247,77 @@ void (*command_handlers[BT_MCS_MEDIA_STATE_LAST])(struct mpl_cmd command,
+ * If found, return pointer to the group
+ * otherwise, the pointer returned is
+ */
+static void find_group_by_id(struct mpl_mediaplayer_t *pl, uint64_t id,
+			     struct mpl_group_t **group)
```
```suggestion
static void find_group_by_id(const struct mpl_mediaplayer_t *pl, uint64_t id,
			     struct mpl_group_t **group)
```

Thalley commented on subsys/bluetooth/host/audio/mpl.c:
```diff
@@ -2204,6 +2247,77 @@ void (*command_handlers[BT_MCS_MEDIA_STATE_LAST])(struct mpl_cmd command,
+{
+	struct mpl_group_t *tmp_group = pl->group;
+
+	while (tmp_group->prev != NULL) {
+		tmp_group = tmp_group->prev;
+	}
```
Ditto

Thalley commented on subsys/bluetooth/host/audio/mpl.c:
```diff
@@ -2301,31 +2415,44 @@ uint64_t current_track_id_get(void)
-	media_proxy_pl_current_track_id_cb(id);
-	media_proxy_pl_track_changed_cb(); /* Todo: Spec says not to notify the client */
-				/* who set the track */
+	find_track_by_id(&pl, id, &group, &track);
+
+	if (group != NULL && track != NULL) {
```
Consider changing the return value of `find_track_by_id` to a boolean, so you can do something like

```suggestion
	if (find_track_by_id(&pl, id, &group, &track)) {
```

asbjornsabo replied:
Good idea!  Changed.

Thalley commented on subsys/bluetooth/host/audio/mpl.c:
```diff
@@ -2301,31 +2415,44 @@ uint64_t current_track_id_get(void)
+
+	BT_DBG("Track not found");
+
+	/* TODO: Should an error be returned here?
+	 * That would require a rewrite of the whole MPL api
+	 */
```
The second sentence needs some more explaining here I think. Why would that require a rewrite of the whole MPL API?

asbjornsabo replied:
It would require the functions to have return values, which they currently do not.
Comment added.

Thalley commented on subsys/bluetooth/host/audio/mpl.c:
```diff
@@ -2301,31 +2415,44 @@ uint64_t current_track_id_get(void)
+		}
+	}
+
+	BT_DBG("Track not found");
```
Isn't this logged always, even if the track is found?

asbjornsabo replied:
Added missing return

Thalley commented on subsys/bluetooth/host/audio/mpl.c:
```diff
@@ -2336,17 +2463,22 @@ uint64_t next_track_id_get(void)
+
+	BT_DBG_OBJ_ID("Next Track ID to set: ", id);
+
+	find_track_by_id(&pl, id, &group, &track);
+
+	if (group != NULL && track != NULL) {
```
Ditto

asbjornsabo replied:
Changed.

Thalley commented on subsys/bluetooth/host/audio/mpl.c:
```diff
@@ -2356,10 +2488,28 @@ uint64_t current_group_id_get(void)
-	pl.group->id = id;
-	media_proxy_pl_current_group_id_cb(pl.group->id);
+
+	find_group_by_id(&pl, id, &group);
+
+	if (group != NULL) {
```
Same comment w.r.t. the return value

asbjornsabo replied:
Changed

Thalley commented on subsys/bluetooth/host/audio/mpl.c:
```diff
@@ -2356,10 +2488,28 @@ uint64_t current_group_id_get(void)
+		}
+	}
+
+	BT_DBG("Group not found");
```
This looks like it is also always logged

asbjornsabo replied:
Added missing return

Thalley commented on subsys/bluetooth/host/audio/mpl_internal.h:
```diff
@@ -84,6 +84,12 @@ struct mpl_mediaplayer_t {
 	uint8_t              content_ctrl_id;
 	struct media_proxy_pl_calls calls;
+
+	bool                        next_track_set; /* If next track explicitly set */
```
I'm wondering if we should append a "pending" to this field to make it more clear that is a pending action, e.g. `next_track_set_pending`.

asbjornsabo replied:
Not sure if that is clearer.  When the next track has been set, it has been set, and is no longer pending.
There is always (with a few exceptions) a next track, and this one is no more pending than the one that would have been used if it was not set explicitly.

Thalley replied:
I am not sure I follow, or perhaps even understand this field. 
This is some boolean that is set by a client (or the server itself perhaps as well?) to change to the next track. At some point we check this boolean, and if it is set, we change the track and clear it. 

To me that sounds like a short term state variable for a pending action that is cleared once the action has been executed, but perhaps I am misunderstanding something?

asbjornsabo replied:
It is not a boolean signaling to change to the next track.  It is a boolean to signal that the next track has been set to a different track than the one that would be chosen from the normal playing order.  And the struct says which track this is.
So, at the next track change, this (given) track shall become the current track, instead of the current track's "normal" next track.
There is always a next track, but in this case which track is the next track is an exception from the normal.

(This could be solved in other ways as well, but the ones I am thinking of involves a more elaborate group/track pointers, and I do not feel that is worth it now.)

Thalley commented on subsys/bluetooth/host/audio/mpl_internal.h:
```diff
@@ -84,6 +84,12 @@ struct mpl_mediaplayer_t {
+
+	bool                        next_track_set; /* If next track explicitly set */
+	struct {
+		struct mpl_track_t  *track; /* The track explicitly set as next track */
+		struct mpl_group_t  *group; /* The group of the set track */
```
Is it only the track that can be explicitly set, or also the group?

asbjornsabo replied:
For "next track", it is the track that is set.  The group of the track is stored to avoid having to do a "find" again later.

Thalley replied:
I see :) I guess the `group` of the `track` isn't stored in the `mpl_track_t` struct right, since we need this here?

asbjornsabo replied:
Exactly, that is how it is.
As I have commented above, the data structure could be changed so that each track has a pointer to it's group (which would be fun to do, ad would also help with the "set next track"), but I do not think that is worth the effort now.

Thalley commented on subsys/bluetooth/shell/mcc.c:
```diff
@@ -751,6 +799,21 @@ int cmd_mcc_read_current_track_obj_id(const struct shell *sh, size_t argc,
+	int result;
+	uint64_t id = strtoul(argv[1], NULL, 0);
+
+	id = id & 0x0000FFFFFFFFFFFF; /* 48 bits only */
```
```suggestion
	id = id & 0x0000FFFFFFFFFFFFUL; /* 48 bits only */
```

Thalley replied:
Since this is a value that is used multiple places, suggest to add a #define for it.

asbjornsabo replied:
https://github.com/zephyrproject-rtos/zephyr/pull/38528

asbjornsabo replied:
Fixed three places.
#define suggested in upsteream PR.  (Should be in OTS.)

Thalley commented on subsys/bluetooth/shell/mcc.c:
```diff
@@ -763,6 +826,21 @@ int cmd_mcc_read_next_track_obj_id(const struct shell *sh, size_t argc,
+	int result;
+	uint64_t id = strtoul(argv[1], NULL, 0);
+
+	id = id & 0x0000FFFFFFFFFFFF; /* 48 bits only */
```
Ditto

Thalley commented on subsys/bluetooth/shell/mcc.c:
```diff
@@ -775,6 +853,21 @@ int cmd_mcc_read_current_group_obj_id(const struct shell *sh, size_t argc,
+	int result;
+	uint64_t id = strtoul(argv[1], NULL, 0);
+
+	id = id & 0x0000FFFFFFFFFFFF; /* 48 bits only */
```
Ditto

Thalley commented on subsys/bluetooth/shell/mcc.c:
```diff
@@ -1348,12 +1441,21 @@ SHELL_STATIC_SUBCMD_SET_CREATE(mcc_cmds,
 	SHELL_CMD_ARG(read_current_track_obj_id, NULL,
 		      "Read Current Track Object ID",
 		      cmd_mcc_read_current_track_obj_id, 1, 0),
+	SHELL_CMD_ARG(set_current_track_obj_id, NULL,
+		      "Set Current Track Object ID",
+		      cmd_mcc_set_current_track_obj_id, 2, 0),
 	SHELL_CMD_ARG(read_next_track_obj_id, NULL,
 		      "Read Next Track Object ID",
 		      cmd_mcc_read_next_track_obj_id, 1, 0),
+	SHELL_CMD_ARG(set_next_track_obj_id, NULL,
+		      "Set Next Track Object ID",
+		      cmd_mcc_set_next_track_obj_id, 2, 0),
 	SHELL_CMD_ARG(read_current_group_obj_id, NULL,
 		      "Read Current Group Object ID",
 		      cmd_mcc_read_current_group_obj_id, 1, 0),
+	SHELL_CMD_ARG(set_current_group_obj_id, NULL,
+		      "Set Current Group Object ID",
+		      cmd_mcc_set_current_group_obj_id, 2, 0),
```
Considering adding some way of showing that it expects an ID argument in base 10

asbjornsabo replied:
It doesn't.  Seems to work perfectly well with both hex, octal and decimal input.

Thalley replied:
Still missing some information about what the argument is expected to be. See e.g. the `bt.c` shell commands for how arguments are defined in the help string.

asbjornsabo replied:
Got it.  Fixed.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/mcc_test.c:
```diff
@@ -37,6 +37,7 @@ static uint64_t g_next_track_object_id;
 static uint64_t g_current_group_object_id;
 static uint64_t g_parent_group_object_id;
 static uint64_t g_search_results_object_id;
+static uint64_t g_tmp_object_id;
```
Can't this be a local variable? Doesn't look like it needs to be global.

asbjornsabo replied:
It can be local.  (Will be "half-global" anyway, as the test_main function is so large ...)
Changed.

asbjornsabo commented on include/bluetooth/mcc.h:
```diff
@@ -155,6 +155,15 @@ typedef void (*bt_mcc_segments_obj_id_read_cb)(struct bt_conn *conn,
+ *
+ * @param conn          The connection that was used to initialise MCC
+ * @param err           Error value. 0 on success, GATT error or ERRNO on fail
+ * @param obj_id        The Object ID written (or attempted to write) (48 bits)
```
Changed to a third way ...

asbjornsabo commented on include/bluetooth/mcc.h:
```diff
@@ -155,6 +155,15 @@ typedef void (*bt_mcc_segments_obj_id_read_cb)(struct bt_conn *conn,
 
+/** @brief Callback function for mcc_set_current_track_obj_id
+ *
+ * @param conn          The connection that was used to initialise MCC
```
MCC should be OK in the MCC header file?  Especially if defined at the start?

Thalley replied:
At least when upstreaming VCS and MICS I was asked to write e.g. Volume Control Service fully

asbjornsabo replied:
MCC -> "the media control client"

asbjornsabo committed:
Bluetooth: Audio: Media control client cleanup

Various clean ups - function names in descriptions, case, indents,
spell client out in full, ...

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented on include/bluetooth/mcc.h:
```diff
@@ -155,6 +155,15 @@ typedef void (*bt_mcc_segments_obj_id_read_cb)(struct bt_conn *conn,
+/** @brief Callback function for mcc_set_current_track_obj_id
+ *
+ * @param conn          The connection that was used to initialise MCC
+ * @param err           Error value. 0 on success, GATT error or ERRNO on fail
```
Fixed everywhere

asbjornsabo commented on include/bluetooth/mcc.h:
```diff
@@ -500,12 +530,39 @@ int bt_mcc_read_segments_obj_id(struct bt_conn *conn);
+ *
+ * @param obj_id    ID (in OTS) of the track to set as the next track
+ */
+int bt_mcc_set_next_track_obj_id(struct bt_conn *conn, uint64_t obj_id);
```
For consistency reasons, this renaming should be done for all functions (and callbacks) in the file.
It will of course also involve updating all users/tests of the module.
While useful, I am severely tempted to declare this out of scope for this PR.

asbjornsabo committed:
Bluetooth: Audio: Media - fix error code

Return -EINVAL instead of -ENOTCONN if given a NULL connection
pointer.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

Thalley commented on include/bluetooth/mcc.h:
```diff
@@ -500,12 +530,36 @@ int bt_mcc_read_segments_obj_id(struct bt_conn *conn);
 /** @brief Read Current Track Object ID */
 int bt_mcc_read_current_track_obj_id(struct bt_conn *conn);
 
+/** @brief Set Current Track Object ID
+ *
+ * Set the Current Track to the the track given by the @p id parameter
+ *
+ * @param id   Object Transfer Service ID (UINT48) of the track to set as the current track
+ */
+int bt_mcc_set_current_track_obj_id(struct bt_conn *conn, uint64_t id);
+
 /** @brief Read Next Track Object ID */
 int bt_mcc_read_next_track_obj_id(struct bt_conn *conn);
 
+/** @brief Set Next Track Object ID
+ *
+ * Set the Next Track to the the track given by the @p id parameter
+ *
+ * @param id   Object Transfer Service ID (UINT48) of the track to set as the next track
+ */
+int bt_mcc_set_next_track_obj_id(struct bt_conn *conn, uint64_t obj_id);
+
 /** @brief Read Current Group Object ID */
 int bt_mcc_read_current_group_obj_id(struct bt_conn *conn);
 
+/** @brief Set Current Group Object ID
+ *
+ * Set the Current Group to the the group given by the @p id parameter
+ *
+ * @param obj_id   Object Transfer Service ID (UINT48) of the group to set as the current group
+ */
+int bt_mcc_set_current_group_obj_id(struct bt_conn *conn, uint64_t obj_id);
+
```
This is kinda fun:
In `bt_mcc_set_current_track_obj_id` you reference `@p id`, document `id` and the argument is `id` - All is good
In `bt_mcc_set_next_track_obj_id` you reference `@p id`, document `id` but the argument is `obj_id` - One issue here
In `bt_mcc_set_current_group_obj_id` you refernce `@p id`, document `obj_id` and the argument is `obj_id` - One or two issue here. 

A few consistency fixes are needed here

asbjornsabo replied:
I thought I had changed them all, but obviously a few slipped in the hurry yesterday.
Fixed now.
